File: k_means/cs/kmeans.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace KMeansClustering
{
    class Point
    {
        public double X { get; set; }
        public double Y { get; set; }
    }

    class Cluster
    {
        public Point Centroid { get; set; }
        public int NumPoints { get; set; }
    }

    class Program
    {
        private const int K = 3; // Number of clusters
        private const int MaxIterations = 100;

        static double Distance(Point p1, Point p2)
        {
            return Math.Sqrt(Math.Pow(p2.X - p1.X, 2) + Math.Pow(p2.Y - p1.Y, 2));
        }

        static void InitializeCentroids(List<Point> dataPoints, List<Cluster> clusters)
        {
            Random random = new Random();

            for (int i = 0; i < K; i++)
            {
                int randomIndex = random.Next(dataPoints.Count);
                clusters[i].Centroid = dataPoints[randomIndex];
                clusters[i].NumPoints = 0;
            }
        }

        static void AssignPointsToCentroids(List<Point> dataPoints, List<Cluster> clusters)
        {
            foreach (var dataPoint in dataPoints)
            {
                double minDistance = Distance(dataPoint, clusters[0].Centroid);
                int clusterIndex = 0;

                for (int j = 1; j < K; j++)
                {
                    double currDistance = Distance(dataPoint, clusters[j].Centroid);
                    if (currDistance < minDistance)
                    {
                        minDistance = currDistance;
                        clusterIndex = j;
                    }
                }

                clusters[clusterIndex].NumPoints++;
            }
        }

        static void UpdateCentroids(List<Point> dataPoints, List<Cluster> clusters)
        {
            foreach (var cluster in clusters)
            {
                double sumX = 0.0, sumY = 0.0;
                int numPoints = cluster.NumPoints;

                foreach (var dataPoint in dataPoints)
                {
                    sumX += dataPoint.X;
                    sumY += dataPoint.Y;
                }

                cluster.Centroid.X = sumX / numPoints;
                cluster.Centroid.Y = sumY / numPoints;
            }
        }

        static void KMeans(List<Point> dataPoints)
        {
            List<Cluster> clusters = new List<Cluster>(K);

            for (int i = 0; i < K; i++)
            {
                clusters.Add(new Cluster());
            }

            // Step 1: Initialize centroids randomly
            InitializeCentroids(dataPoints, clusters);

            int iteration = 0;
            while (iteration < MaxIterations)
            {
                // Step 2: Assign points to centroids
                AssignPointsToCentroids(dataPoints, clusters);

                // Step 3: Update centroids
                UpdateCentroids(dataPoints, clusters);

                iteration++;
            }

            // Print the final clusters
            for (int i = 0; i < K; i++)
            {
                Console.WriteLine($"Cluster {i + 1}: Centroid ({clusters[i].Centroid.X}, {clusters[i].Centroid.Y}), Points: {clusters[i].NumPoints}");
            }
        }

        static void Main()
        {
            // Sample data points
            List<Point> dataPoints = new List<Point>
            {
                new Point { X = 2.0, Y = 3.0 },
                new Point { X = 2.5, Y = 4.5 },
                new Point { X = 1.5, Y = 2.5 },
                new Point { X = 6.0, Y = 5.0 },
                new Point { X = 7.0, Y = 7.0 },
                new Point { X = 5.0, Y = 5.5 },
                new Point { X = 9.0, Y = 2.0 },
                new Point { X = 10.0, Y = 3.5 },
                new Point { X = 9.5, Y = 2.5 }
            };

            // Perform K-means clustering
            KMeans(dataPoints);
        }
    }
}

File: k_means/c/kmeans.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define K 3 // Number of clusters
#define MAX_ITERATIONS 100

typedef struct {
    double x;
    double y;
} Point;

typedef struct {
    Point centroid;
    int numPoints;
} Cluster;

// Function to calculate Euclidean distance between two points
double distance(Point p1, Point p2) {
    return sqrt(pow(p2.x - p1.x, 2) + pow(p2.y - p1.y, 2));
}

// Function to initialize centroids randomly
void initializeCentroids(Point* dataPoints, Cluster* clusters, int numDataPoints) {
    for (int i = 0; i < K; i++) {
        int randomIndex = rand() % numDataPoints;
        clusters[i].centroid = dataPoints[randomIndex];
        clusters[i].numPoints = 0;
    }
}

// Function to assign each data point to the nearest centroid
void assignPointsToCentroids(Point* dataPoints, Cluster* clusters, int numDataPoints) {
    for (int i = 0; i < numDataPoints; i++) {
        double minDistance = distance(dataPoints[i], clusters[0].centroid);
        int clusterIndex = 0;

        for (int j = 1; j < K; j++) {
            double currDistance = distance(dataPoints[i], clusters[j].centroid);
            if (currDistance < minDistance) {
                minDistance = currDistance;
                clusterIndex = j;
            }
        }

        clusters[clusterIndex].numPoints++;
    }
}

// Function to update the centroids based on the assigned points
void updateCentroids(Point* dataPoints, Cluster* clusters) {
    for (int i = 0; i < K; i++) {
        double sumX = 0.0, sumY = 0.0;
        int numPoints = clusters[i].numPoints;

        for (int j = 0; j < numPoints; j++) {
            sumX += dataPoints[j].x;
            sumY += dataPoints[j].y;
        }

        clusters[i].centroid.x = sumX / numPoints;
        clusters[i].centroid.y = sumY / numPoints;
    }
}

// Function to perform K-means clustering
void kMeans(Point* dataPoints, int numDataPoints) {
    Cluster clusters[K];

    // Step 1: Initialize centroids randomly
    initializeCentroids(dataPoints, clusters, numDataPoints);

    int iteration = 0;
    while (iteration < MAX_ITERATIONS) {
        // Step 2: Assign points to centroids
        assignPointsToCentroids(dataPoints, clusters, numDataPoints);

        // Step 3: Update centroids
        updateCentroids(dataPoints, clusters);

        iteration++;
    }

    // Print the final clusters
    for (int i = 0; i < K; i++) {
        printf("Cluster %d: Centroid (%.2f, %.2f), Points: %d\n",
            i + 1, clusters[i].centroid.x, clusters[i].centroid.y, clusters[i].numPoints);
    }
}

int main() {
    // Sample data points
    Point dataPoints[] = {
        {2.0, 3.0},
        {2.5, 4.5},
        {1.5, 2.5},
        {6.0, 5.0},
        {7.0, 7.0},
        {5.0, 5.5},
        {9.0, 2.0},
        {10.0, 3.5},
        {9.5, 2.5}
    };

    int numDataPoints = sizeof(dataPoints) / sizeof(dataPoints[0]);

    // Perform K-means clustering
    kMeans(dataPoints, numDataPoints);

    return 0;
}